modin/core/execution/ray/common/deferred_execution.py

```diff
@@ -0,0 +1,707 @@
+    """
+    Execution exception wrapper.
+
+    Parameters
+    ----------
+    msg : str
+    cause : Exception
+    """
+
+    def __init__(self, msg: str, cause: Exception):
+        super().__init__(msg, cause)
+
+
+class _Tag(Enum):  # noqa: PR01
+    """A set of special values used for the method arguments de/construction."""
+
+    # The next item is an execution chain
+    CHAIN = 0
+    # The next item is a reference
+    REF = 1
+    # The next item a list
+    LIST = 2
+    # End of list or chain
+    END = 3
+
+
+ListOrDe = (DeferredExecution, list, tuple)
+
+
+def has_list_or_de(it: Iterable):
+    """
+    Check if the specified iterable contains either a list or a DeferredExecution object.
+
+    Parameters
+    ----------
+    it : Iterable
+
+    Returns
+    -------
+    bool
+    """
+    return any(isinstance(i, ListOrDe) for i in it)
+
+
+def deconstruct(de: DeferredExecution) -> Tuple[List[DeferredExecution], List[Any]]:
+    """
+    Convert the specified execution tree to a flat list.
+
+    This is required for the automatic Ray object references
+    materialization before passing the list to a Ray worker.
+
+    The format of the list is the following:
+    <input object> sequence<<function> <n><args> <n><kwargs> <ref> <res>>...
+    If <n> is >= 0, then the next n objects are the function arguments.
+    If <n> is -1, it means that the method arguments contain list and/or
+    DeferredExecution (chain) objects. In this case the next values are read
+    one by one until `_Tag.END` is encountered. If the value is `_Tag.LIST`,
+    then the next sequence of values up to `_Tag.END` is converted to list.
+    If the value is `_Tag.CHAIN`, then the next sequence of values up to
+    `_Tag.END` has exactly the same format, as described here.
+    If the value is `_Tag.REF`, then the next value is a reference id, i.e.
+    the actual value should be retrieved by this id from the previously
+    saved objects. The <input object> could also be `_Tag.REF` or `_Tag.LIST`.
+
+    If <n> before <kwargs> is >=0, then the next 2*n values are the argument
+    names and values in the following format - [name1, value1, name2, value2...].
+    If it's -1, then the next values are converted to list in the same way as
+    <args> and the argument names are the next len(<args>) values.
+
+    <ref> is an integer reference id. If it's not 0, then there is another
+    chain referring to the execution result of this method and, thus, it must
+    be saved so that other chains could retrieve the object by the id.
+
+    <res> is a 'get result' flag. If it's True, then the method execution
+    result must not only be passed to the next method in the chain, but also
+    returned to the caller. The object length and width are added to the meta list.
+
+    Parameters
+    ----------
+    de : DeferredExecution
+
+    Returns
+    -------
+    list
+    """
+    stack = []
+    result_consumers = []
+    output = []
+    de.ref_count(1)
```

Comment:
Why do we increment ref count here?

Reply:
Since we start execution from this node, we need to get the result for this node. When traversing the tree up, we dereference each parent node and decrement the counter. If the counter is 0, the result for this node is not returned.